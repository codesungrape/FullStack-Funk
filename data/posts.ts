// data/posts.ts

// You might want to add a type for your posts
export type Post = {
    title: string
    excerpt: string
    author: {
      name: string
      avatar: string
    }
    date: string
    readTime: string
    tags: string[]
    image: string
    slug: string
  }
  
  export const posts: Post[] = [
    {
      title: "Understanding React Server Components",
      excerpt: "Learn how React Server Components work and how they can improve your application's performance...",
      author: {
        name: "Sarah Johnson",
        avatar: "/placeholder.svg?height=40&width=40",
      },
      date: "Mar 15, 2024",
      readTime: "5 min",
      tags: ["React", "JavaScript", "Web Development"],
      image: "/placeholder.svg?height=400&width=800",
      slug: "understanding-react-server-components",
    },
    {
        title: "CSS Grid: A Complete Guide",
        excerpt: "Master CSS Grid layout with this comprehensive guide covering all the essential concepts...",
        author: {
          name: "Mike Chen",
          avatar: "/placeholder.svg?height=40&width=40",
        },
        date: "Mar 12, 2024",
        readTime: "8 min",
        tags: ["CSS", "Web Development", "Tutorial"],
        image: "/placeholder.svg?height=400&width=800",
        slug: "css-grid-complete-guide",
      },
      {
        title: "TypeScript Best Practices in 2024",
        excerpt: "Discover the latest TypeScript best practices and patterns to write better, more maintainable code...",
        author: {
          name: "Emma Wilson",
          avatar: "/placeholder.svg?height=40&width=40",
        },
        date: "Mar 10, 2024",
        readTime: "6 min",
        tags: ["TypeScript", "JavaScript", "Best Practices"],
        image: "/placeholder.svg?height=400&width=800",
        slug: "typescript-best-practices-2024",
      },
    
  ]