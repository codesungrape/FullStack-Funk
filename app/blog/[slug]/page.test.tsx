import { render, screen } from '@testing-library/react'
import BlogPost, { generateMetadata, generateStaticParams } from './page'
import { posts } from '@/data/posts'
import { notFound } from 'next/navigation'

// Mock next/navigation
jest.mock('next/navigation', () => ({
  notFound: jest.fn(() => {
    throw new Error('NEXT_NOT_FOUND')
  })
}))

// Mock next/image
jest.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, className }: any) => (
    <img src={src} alt={alt} className={className} />
  )
}))

// Mock Layout component
jest.mock('@/components/page-layout/page-layout', () => ({
  __esModule: true,
  default: ({ children }: { children: React.ReactNode }) => <div>{children}</div>
}))

// Mock post data
const mockPost = {
  slug: 'test-post',
  title: 'Test Post',
  image: '/test-image.jpg',
  author: {
    name: 'John Doe',
    avatar: '/avatar.jpg'
  },
  date: '2024-02-20',
  readTime: '5 min read',
  tags: ['test', 'blog'],
  excerpt: 'This is a test post excerpt'
}

describe('BlogPost', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })

  it('renders the blog post when slug is found', () => {
    // Mock posts array to include our test post
    jest.spyOn(posts, 'find').mockReturnValue(mockPost)

    render(<BlogPost params={{ slug: 'test-post' }} />)

    // Check if main elements are rendered
    expect(screen.getByRole('article')).toBeInTheDocument()
    expect(screen.getByRole('heading', { name: mockPost.title })).toBeInTheDocument()
    expect(screen.getByText(mockPost.author.name)).toBeInTheDocument()
    expect(screen.getByText(mockPost.date)).toBeInTheDocument()
    expect(screen.getByText(mockPost.readTime)).toBeInTheDocument()
    expect(screen.getByText(mockPost.excerpt)).toBeInTheDocument()

    // Check if tags are rendered
    mockPost.tags.forEach(tag => {
      expect(screen.getByText(tag)).toBeInTheDocument()
    })

    // Check if images are rendered with correct props
    const coverImage = screen.getByAltText(mockPost.title)
    expect(coverImage).toBeInTheDocument()
    expect(coverImage).toHaveAttribute('src', mockPost.image)

    const authorAvatar = screen.getByAltText(mockPost.author.name)
    expect(authorAvatar).toBeInTheDocument()
    expect(authorAvatar).toHaveAttribute('src', mockPost.author.avatar)
  })

  it('calls notFound when slug is not found', () => {
    // Mock posts array to return undefined (post not found)
    jest.spyOn(posts, 'find').mockReturnValue(undefined)

    // Expect the render to throw because notFound throws
    expect(() => {
      render(<BlogPost params={{ slug: 'non-existent-post' }} />)
    }).toThrow('NEXT_NOT_FOUND')

    // Check if notFound was called
    expect(notFound).toHaveBeenCalled()
  })

   it('renders with correct CSS classes', () => {
      jest.spyOn(posts, 'find').mockReturnValue(mockPost)
  
      const { container } = render(<BlogPost params={{ slug: 'test-post' }} />)
  
      // Check if elements have the correct CSS classes
      expect(container.querySelector('.article')).toBeInTheDocument()
      expect(container.querySelector('.coverImage')).toBeInTheDocument()
      expect(container.querySelector('.content')).toBeInTheDocument()
      expect(container.querySelector('.metadata')).toBeInTheDocument()
      expect(container.querySelector('.author')).toBeInTheDocument()
      expect(container.querySelector('.tags')).toBeInTheDocument()
      expect(container.querySelector('.tag')).toBeInTheDocument()
    })
})

  // Snapshot test
  it('matches snapshot', () => {
    jest.spyOn(posts, 'find').mockReturnValue(mockPost)
    
    const { container } = render(<BlogPost params={{ slug: 'test-post' }} />)
    expect(container).toMatchSnapshot()
  })

  describe('generateMetadata', () => {
    it('returns correct metadata for existing post', () => {
      jest.spyOn(posts, 'find').mockReturnValue(mockPost)
  
      const metadata = generateMetadata({ params: { slug: 'test-post' } })
  
      expect(metadata).toEqual({
        title: mockPost.title,
        description: mockPost.excerpt
      })
    })
  
    it('returns not found metadata for non-existent post', () => {
      jest.spyOn(posts, 'find').mockReturnValue(undefined)
  
      const metadata = generateMetadata({ params: { slug: 'non-existent' } })
  
      expect(metadata).toEqual({
        title: 'Post Not Found',
        description: 'The requested blog post could not be found'
      })
    })
  })
  
  describe('generateStaticParams', () => {
    it('returns correct params for all posts', () => {
      const mockPosts = [
        { ...mockPost, slug: 'post-1' },
        { ...mockPost, slug: 'post-2' }
      ]
      
      // Mock the entire posts array
      jest.spyOn(posts, 'map').mockImplementation(() => 
        mockPosts.map(post => ({ slug: post.slug }))
      )
  
      const params = generateStaticParams()
  
      expect(params).toEqual([
        { slug: 'post-1' },
        { slug: 'post-2' }
      ])
    })
  })