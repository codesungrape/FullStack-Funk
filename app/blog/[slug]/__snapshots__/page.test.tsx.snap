// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div>
    <article
      class="article"
    >
      <img
        alt="Test Post"
        class="coverImage"
        src="/test-image.jpg"
      />
      <div
        class="content"
      >
        <h1>
          Test Post
        </h1>
        <div
          class="metadata"
        >
          <div
            class="author"
          >
            <img
              alt="John Doe"
              src="/avatar.jpg"
            />
            <span>
              John Doe
            </span>
          </div>
          <time>
            2024-02-20
          </time>
          <span>
            5 min read
          </span>
        </div>
        <div
          class="tags"
        >
          <span
            class="tag"
          >
            test
          </span>
          <span
            class="tag"
          >
            blog
          </span>
        </div>
        <p>
          This is a test post excerpt
        </p>
      </div>
    </article>
  </div>
</div>
`;
